What you will learn:
- How to add, subtract, multiply divide numbers
- Float division and integer division
- fstring
- Modulus (finding the remainder)
- Exponents
- Exceptions (ZeroDivisionError, NameError, SyntaxError, TypeError)
- Adding wrong types
- Trackback and exception print out to terminal

Okay now lets do more cool things with variables. Like making python do math for us!

Below are simple examples of how we can add and subtract. Note here that we are adding and subtracting ints and floats in the same expression.

In [5]:
a = 3
b = 1

x = a + b
print(x)

a = 3
b = 1.5

y = a - b
print(y)

4
1.5


This is fairly striaght forward so let's discuss how we can print information in a more readable way using fstring

In [6]:
age = 22
print(22)  # this is find but what happens when we have lots of varables being printed. Let's clean this up

print(f"Hello I am {age} years old")

name = "Bob"

print(f"{name}, did you just turn {age}?")

22
Hello I am 22 years old
Bob, did you just turn 22?


Division can be broken down into 3 types: integer, float, and modulus.

In [7]:
# Float
pi = 3.14159
x = 2

a = pi / x

print(f"Float: pi/2 = {a}")

# Integer

b = pi // x

print(f"Integer: pi/2 = {b}")


# Modulus

c = pi % x

print(f"Modulus: pi/2 = {c}")

Float: pi/2 = 1.570795
Integer: pi/2 = 1.0
Modulus: pi/2 = 1.1415899999999999


Looking at the abvoe output ...

Float is the "normal" division you are used to and will return a rational number.

Integer is how many time the denominator exactly into the numerator and does not account for a remainder. So 2 fits into 3.14159 one time (3.14159 - 2 = 1.14159); 2 does not fit into 1.14159 so the result is 1.

Modulus shows the remainder. pi/2 could be written as 1 remainder 1.14159.

We can also add strings together

In [8]:
intro = "My name is Bob. "
funFacts = "I like to play the drums. "
age = "Yesterday I turned 20!"

personalInfo = intro + funFacts + age

print(personalInfo)

My name is Bob. I like to play the drums. Yesterday I turned 20!


Now lets have a look at exponents

In [9]:
x = 2 ** 8

print(f"An 8-bit LED can reprsent {x} color combinations")

km = 127

m = km * 10 ** 3

print(f"{km}km is the same as {m}m")
print(f"\nWe can also add commas when printing large numbers like this:\n {km}km is the same as {m:,}m")

An 8-bit LED can reprsent 256 color combinations
127km is the same as 127000m

We can also add commas when printing large numbers like this:
 127km is the same as 127,000m


Okay now let's talk about how to break your code. What happens in your calulator when you divide by 0? What about in code?

In [10]:
x = 5
y = 0

z = x/y

ZeroDivisionError: division by zero

Python will throw a ZeroDivisionError. This will cause the program to crash. 

Notive the output that comes when this happens. There is a section that says "Traceback (most recent call last)". This is telling us the last part of the code that ran before the crash occured. As programs get larger and more complicated this traceback is invaluable to determine where an error occurs. See here that we are told that the error ocurred on line 4 and the error type was a ZeroDivisionError.

What are some other errors?

In [11]:
grade1 = 88
grade2 = 87
grade3 = 71
grade4 = 84

gradeSum = grade1 + grade2 + grade3 + grade4

average = gradeSum / numOfGrades

print(f"Average grade was: {average}")

NameError: name 'numOfGrades' is not defined

NameError is thrown when a varable is referenced but is not defined.

Fix the code above so that is works. See the Traceback to find the line number where the error was thrown. 

NameError is when a varble is eferenced but is not defined. What happens when a varible is simply defined wrong?

In [12]:
string = "This is a string

SyntaxError: EOL while scanning string literal (<ipython-input-12-c2a4f9d9293f>, line 1)

In [13]:
age 31

SyntaxError: invalid syntax (<ipython-input-13-89df32bfd07f>, line 1)

A SyntaxError occurs when the python lanaugage is used wrong. Pyton doesn't know what you are trying to do and so the error is thrown

What happens when varable types are operated on in ways they cannot handle. Such as adding a string to an int

In [14]:
num = 5
string = "Test "

z = string + num

TypeError: can only concatenate str (not "int") to str

Note: The structure of these programs is very open. As we move forward the design requirments of your programs will become more rigid. For now, if you could use the program to anwser the question that is good enough.

What you need to do:
Pt 1
- solve x = 45 + 4(4/6+2)^5 --- should = 584.391
- solve x = (4/5*5+1)^.5    --- should = 2.23607
Pt 2
- write code to determine if 39879827239498734985798 is divisible by 3 without a remainder
- write a program that can solve if any given number is even or not